src/psyclone/domain/lfric/psyir.py

```diff
@@ -0,0 +1,162 @@
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford STFC Daresbury Lab
+
+'''This module generates LFRic-specific PSyIR classes from a list of
+definitions.
+
+'''
+# pylint: disable=unused-import
+# pylint: disable=exec-used
+from psyclone.psyir.symbols import ContainerSymbol, DataSymbol, \
+    DeferredType, GlobalInterface, ScalarType, ArrayType
+
+# Define LFRic symbols and datatypes
+modules = [{"name": "constants_mod", "vars": ["i_def", "r_def", "l_def"]}]
+
+# Generate LFRic module symbols from definitions
+for module in modules:
+    MODULE_NAME = module["name"]
+    exec("{0} = ContainerSymbol('{1}')\n".format(
+        MODULE_NAME.upper(), MODULE_NAME))
+    for module_var in module["vars"]:
+        exec("{0} = DataSymbol('{1}', DeferredType(), interface="
+             "GlobalInterface({2}))".format(
+                 module_var.upper(), module_var, MODULE_NAME.upper()))
+
+scalar_datatypes = [
+    {"name": "lfric integer scalar", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "integer_scalar"},
+    {"name": "lfric real scalar", "intrinsic": "real",
+     "precision": "r_def", "vars": [], "short_name": "real_scalar"},
+    {"name": "lfric logical scalar", "intrinsic": "boolean",
+     "precision": "l_def", "vars": [], "short_name": "logical_scalar"},
+    {"name": "cell position", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nlayers"},
+    {"name": "mesh height", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nlayers"},
+    {"name": "operator size", "intrinsic": "integer",
```

Comment:
I don't think that ncell_3d should be called operator size - it's the total number of cells in the (local) partition. We should also now make sure it doesn't get passed to a given kernel more than once.

Reply:
I've changed the name but there may still be a better one.

I'm not sure about the arg not being passed in multiple times. That's a change to the API isn't it, which would affect existing LFRic kernels and we would also need to decide where to add it. Sounds like a separate PR to me :-)